mis_builder_tax_report/models/mis_builder_tax_report.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) ACSONE SA/NV 2022
+
+from typing import Iterator
+
+from odoo import _, api, models
+from odoo.exceptions import UserError
+
+
+class MisTaxReportBuilder(models.AbstractModel):
+    _name = "mis.tax.report.builder"
+
+    @api.model
+    def _iter_account_tax_report_lines(
+        self, lines: models.Model, level: int = 0
+    ) -> Iterator[models.Model]:
+        """Iterate account.tax.report.line, depth first."""
+        for line in lines:
+            yield line, level
+            yield from self._iter_account_tax_report_lines(
+                line.children_line_ids, level + 1
+            )
+
+    @api.model
+    def _xmlid_from_object(self, o: models.Model) -> str:
+        imd = (
+            self.env["ir.model.data"]
+            .sudo()
+            .search([("model", "=", o._name), ("res_id", "=", o.id)], limit=1)
+        )
+        if not imd:
+            return None
+        return f"{imd.module}.{imd.name}"
+
+    @api.model
+    def _kpi_name_from_tax_report_line(self, line: models.Model) -> str:
+        if line.code:
+            return line.code
+        xmlid = self._xmlid_from_object(line)
+        if xmlid:
+            return xmlid.replace(".", "_")
+        return f"_{line.id}"
+
+    @api.model
+    def _populate_mis_report_template_from_tax_report(
+        self, mis_report_template: models.Model, tax_report: models.Model
+    ) -> None:
+        """Populate the KPIs of a MIS Report template from an Odoo Tax Report.
+
+        TODO
+
+        * Instead of replacing all KPIs, update KPIs with the same name
+          and remove those that do not exist any more in the Odoo report.
+        * Link KPIs to predefined styles based on the level (styles to be added
+          as data in this module).
+        """
+        mis_report_template.kpi_ids.unlink()
+        for sequence, (report_line, _level) in enumerate(
+            self._iter_account_tax_report_lines(tax_report.root_line_ids)
+        ):
+            kpi = self.env["mis.report.kpi"].new()
+            kpi.name = self._kpi_name_from_tax_report_line(report_line)
+            kpi.description = report_line.name
+            kpi.sequence = sequence
+            if report_line.tag_ids and report_line.formula:
+                raise UserError(
+                    _(
+                        "Report line '{name}' has both Tags and a Forumla. "
+                        "We don't grok this."
+                    ).format(name=report_line.name),
+                )
+            if report_line.tag_name:
+                # If the report line has tags, make an expression that gets the balance
+                # of move lines with these tags over the period.
+                kpi.expression = (
+                    f'-balp[][("tax_tag_ids.name", "=", "+{report_line.tag_name}")]'
+                    f'+balp[][("tax_tag_ids.name", "=", "-{report_line.tag_name}")]'
+                )
```

Comment:
Note to self and interested parties. This formula is too simplistic for some localization. In v14 one must take into account the type of entry and possibly also the journal. In v15 this rule is computed into a tax_tag_invert field on account.move.line that also influences the sign to be applied in the report.

So this is to be used with caution and careful testing.

Reply:
@vdewulf